nixos/modules/services/x11/redshift.nix

```diff
@@ -88,51 +71,101 @@ in {
       '';
     };
 
-    extraOptions = mkOption {
-      type = types.listOf types.str;
-      default = [];
-      example = [ "-v" "-m randr" ];
+    settings = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+      };
+
+      default = {};
+
+      example = {
+        redshift = {
+          temp-day = 5500;
+          temp-night = 3700;
+          brightness-day = 1.0;
+          brightness-night = 0.8;
+          fade = 1;
+        };
+        randr = {
+          screen = 0;
+        };
+      };
+
       description = ''
-        Additional command-line arguments to pass to
-        <command>redshift</command>.
+        The configuration to pass to Redshift/Gammastep.
+
+        Available options for Redshift described in
+        <citerefentry>
+          <refentrytitle>redshift</refentrytitle>
+          <manvolnum>1</manvolnum>
+        </citerefentry> and for Gammastep in
+        <citerefentry>
+          <refentrytitle>gammastep</refentrytitle>
+          <manvolnum>1</manvolnum>
+        </citerefentry>.
       '';
     };
   };
 
-  config = mkIf cfg.enable {
+  config = let
+    mainSection = if strings.hasInfix "gammastep" cfg.executable
+      # https://gitlab.com/chinstrap/gammastep/-/commit/1608ed61154cc652b087e85c4ce6125643e76e2f
+      then "general"
+      else "redshift";
+    target = if (cfg.settings.${mainSection}.adjustment-method or null) == "drm"
+      then "basic.target"
+      else "graphical-session.target";
+    configFile = settingsFormat.generate "redshift.conf" cfg.settings;
+    fullExecutable = "${cfg.package}${cfg.executable}";
+  in mkIf cfg.enable {
     # needed so that .desktop files are installed, which geoclue cares about
     environment.systemPackages = [ cfg.package ];
 
-    services.geoclue2.appConfig.redshift = {
+    services.geoclue2.appConfig.redshift = mkIf isGeoclue2 {
       isAllowed = true;
       isSystem = true;
     };
 
-    systemd.user.services.redshift =
-    let
-      providerString = if lcfg.provider == "manual"
-        then "${toString lcfg.latitude}:${toString lcfg.longitude}"
-        else lcfg.provider;
-    in
-    {
+    assertions = [
+      {
+        assertion = (cfg.settings ? "redshift") != (cfg.settings ? "general");
+        message = ''Both [redshift] and [general] sections found in settings. For gammastep use "general = {}" instead of "redshift = {}" in "server.redshift.settings".'';
+      }
+      {
+        assertion = (pathExists fullExecutable);
+        message = ''Executable not found. Are you using "/bin/redshift" for "pkgs.redshift" or "/bin/gammastep" for "pkgs.gammastep"?'';
+      }
+      {
+        assertion = (cfg.settings ? ${mainSection}.dawn-time) == (cfg.settings ? ${mainSection}.dusk-time);
+        message = "Time of dawn and time of dusk must be provided together.";
+      }
+    ];
+
+    services.redshift.settings = {
+      ${mainSection}.location-provider = lcfg.provider;
```

Comment:
This however should use the `redshift` name, since that will be copied to both sections.